### DISCUSSION TITLE
Yolo output to array ?

### DISCUSSION THREAD

**Discussion by Samuel-Bachorik** [ultralytics/yolov5#2412] [General]
Hello I am using this code for detection on image
`
import torch
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True)
img = 'C:/Users/Samuel/PycharmProjects/Condapytorch/image for yolo/KKT.jpg'

results = model(img)
results.print()  # or .show(), .save()
`

I want to ask you is it possible to get this result as array in python ? I mean output image with detected objects as array on tensor.

Or my question is, is there any other way to get array as output ? I want to sent image to yolo and I want it to return output image as array. 

Thank you!

**Comment by glenn-jocher** [comment]
@Samuel-Bachorik yes of course, see the PyTorch Hub Tutorial for examples of returning arrays:

## Tutorials

* [Train Custom Data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data)&nbsp; 🚀 RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://github.com/ultralytics/yolov5/issues/475)
* [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)&nbsp; ⭐ NEW
* [ONNX and TorchScript Export](https://github.com/ultralytics/yolov5/issues/251)
* [Test-Time Augmentation (TTA)](https://github.com/ultralytics/yolov5/issues/303)
* [Model Ensembling](https://github.com/ultralytics/yolov5/issues/318)
* [Model Pruning/Sparsity](https://github.com/ultralytics/yolov5/issues/304)
* [Hyperparameter Evolution](https://github.com/ultralytics/yolov5/issues/607)
* [Transfer Learning with Frozen Layers](https://github.com/ultralytics/yolov5/issues/1314)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

### Detailed Example

To load YOLOv5 from PyTorch Hub for **batched inference** with PIL, OpenCV, Numpy or PyTorch inputs, with results plotted, saved and printed to screen:
```python
import cv2
import torch
from PIL import Image

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True)  # for file/URI/PIL/cv2/np inputs and NMS

# Images
for f in ['zidane.jpg', 'bus.jpg']:  # download 2 images
    print(f'Downloading {f}...')
    torch.hub.download_url_to_file('https://github.com/ultralytics/yolov5/releases/download/v1.0/' + f, f)
img1 = Image.open('zidane.jpg')  # PIL image
img2 = cv2.imread('bus.jpg')[:, :, ::-1]  # OpenCV image (BGR to RGB)
imgs = [img1, img2]  # batched list of images

# Inference
results = model(imgs, size=640)  # includes NMS

# Results
results.print()  
results.save()  # or .show()

# Data
print(results.xyxy[0])  # print img1 predictions (pixels)
#                   x1           y1           x2           y2   confidence        class
# tensor([[7.50637e+02, 4.37279e+01, 1.15887e+03, 7.08682e+02, 8.18137e-01, 0.00000e+00],
#         [9.33597e+01, 2.07387e+02, 1.04737e+03, 7.10224e+02, 5.78011e-01, 0.00000e+00],
#         [4.24503e+02, 4.29092e+02, 5.16300e+02, 7.16425e+02, 5.68713e-01, 2.70000e+01]])
```
<img src="https://user-images.githubusercontent.com/26833433/96368758-114c3a00-1156-11eb-9ede-a03f9aff42e4.jpg" width="480">  <img src="https://user-images.githubusercontent.com/26833433/97121810-54dc1080-1721-11eb-8c25-451ccb1452e4.jpg" width="320">

For all inference options see YOLOv5 `autoShape()` forward method:
https://github.com/ultralytics/yolov5/blob/3551b072b366989b82b3777c63ea485a99e0bf90/models/common.py#L182-L191

### Inference Settings
NMS IoU threshold and confidence threshold are model attributes, and can be modified by:
```python
model.conf = 0.25  # confidence threshold (0-1)
model.iou = 0.45  # NMS IoU threshold (0-1)

results = model(imgs, size=320)  # custom inference size
```

**Reply by pravastacaraka** [reply]
Can I use PyTorch Hub with a custom model?

**Reply by glenn-jocher** [reply]
@pravastacaraka yes of course. See PyTorch Hub Tutorial for details:

## Tutorials

* [Train Custom Data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data)&nbsp; 🚀 RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://github.com/ultralytics/yolov5/issues/475)
* [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)&nbsp; ⭐ NEW
* [ONNX and TorchScript Export](https://github.com/ultralytics/yolov5/issues/251)
* [Test-Time Augmentation (TTA)](https://github.com/ultralytics/yolov5/issues/303)
* [Model Ensembling](https://github.com/ultralytics/yolov5/issues/318)
* [Model Pruning/Sparsity](https://github.com/ultralytics/yolov5/issues/304)
* [Hyperparameter Evolution](https://github.com/ultralytics/yolov5/issues/607)
* [Transfer Learning with Frozen Layers](https://github.com/ultralytics/yolov5/issues/1314)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by Samuel-Bachorik** [comment]
@glenn-jocher  Hello thank you for your reply. I understand basics of pytorch that's no problem. But the problem is I need output image as array. When I use results.save() I want this image that is saved as array. Is this possible ? results.save() saves me output image to my PC but I want this image to array.

**Reply by glenn-jocher** [reply]
@Samuel-Bachorik the results object contains the original images as results.imgs. You can overlay results on the original images with the results.render() method. See detections class for full details:

https://github.com/ultralytics/yolov5/blob/886f1c03d839575afecb059accf74296fad395b6/models/common.py#L232-L247
